### PR TITLE
cmake: Switch to libsecp256k1 upstream build system

### PR DESCRIPTION
**Pull request by hebasto**
To test this PR, one can compare the ouputs of the following commands:
- Autotools:
```
$ make -C src/secp256k1 libsecp256k1.la V=1
```
- CMake:
```
$ cmake --build build -t secp256k1 --verbose
```

Also sets of the libsecp256k1-specific tests can be compared between `make check` and `ctest`.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
 
+option(SECP256K1_DISABLE_SHARED "" ON)
+option(SECP256K1_ENABLE_MODULE_ECDH "" OFF)
+option(SECP256K1_ENABLE_MODULE_RECOVERY "" ON)
+option(SECP256K1_BUILD_BENCHMARK "" OFF)
+option(SECP256K1_BUILD_TESTS "" OFF)
+option(SECP256K1_BUILD_EXHAUSTIVE_TESTS "" OFF)
+option(SECP256K1_BUILD_CTIME_TESTS "" OFF)
+add_subdirectory(secp256k1)
+target_link_libraries(secp256k1 PRIVATE core_base_interface)
```
Comment (hebasto) [inline src/CMakeLists.txt:16]: Flags from the `core_base_interface` won't be printed in "secp256k1 configure summary".

To achieve this, the `SECP256K1_LATE_CFLAGS` variable have to be used and `SECP256K1_LATE_LDFLAGS` one have to be added to the upstream code.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
 include(GNUInstallDirs)
 include(AddWindowsResources)
 
+option(SECP256K1_DISABLE_SHARED "" ON)
+option(SECP256K1_ENABLE_MODULE_ECDH "" OFF)
+option(SECP256K1_ENABLE_MODULE_RECOVERY "" ON)
+option(SECP256K1_BUILD_BENCHMARK "" OFF)
+option(SECP256K1_BUILD_TESTS "" OFF)
+option(SECP256K1_BUILD_EXHAUSTIVE_TESTS "" OFF)
+option(SECP256K1_BUILD_CTIME_TESTS "" OFF)
```
Comment (hebasto) [inline src/CMakeLists.txt:14]: These cache variables might be disabled upstream when building as a subproject.

**Comment by hebasto** [comment]
For CMake-based projects, it is not common to print the summary at the end of the configuration stage.

Printing summaries by subprojects--libsecp256k1 in this PR and minisketch in the future--makes them less useful for the user due to their verbosity.

**Comment by hebasto** [comment]
cc @real-or-random @theuni @laanwj @fanquake

**Comment by fanquake** [comment]
> For CMake-based projects, it is not common to print the summary at the end of the configuration stage.

I'd say that atleast half of the projects we are (going to) build in depends with CMake print some sort of configuration? So I don't think it's uncommon.

> Printing summaries by subprojects--libsecp256k1 in this PR and minisketch in the future--makes them less useful for the user due to their verbosity.

I'm not exactly sure what you're trying to say here. It's not less useful if you actually want to see how things have been configured / are going to be compiled, which, if the plan is to use the CMake build systems of our subtrees, we definitely do. This is useful for developers, in CI etc etc.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
+add_subdirectory(secp256k1)
```
Comment (theuni) [inline src/CMakeLists.txt:15]: Does this need `EXCLUDE_FROM_ALL` to keep it from installing?

Comment (hebasto) [reply src/CMakeLists.txt:15]: Nope :)

Thanks to: https://github.com/hebasto/bitcoin/blob/43a66c55ec8770cf7c21112aac9b997f3f2fb704/src/secp256k1/CMakeLists.txt#L52

Comment (theuni) [reply src/CMakeLists.txt:15]: I think it still makes sense to go ahead and set that option though, since this is what it's meant for. We could remove the `SECP256K1_BUILD_*` options and not have to worry about any new binaries being added in the future.

Comment (hebasto) [reply src/CMakeLists.txt]: Thanks! Reworked.

File: src/CMakeLists.txt (hunk 2 above)
Comment (hebasto) [reply src/CMakeLists.txt]: Irrelevant since https://github.com/hebasto/bitcoin/pull/192#discussion_r1594428809.

**Review by real-or-random — COMMENTED** [review]
Concept ACK 

The libsecp256k1 config looks good to me.

**Comment by hebasto** [comment]
Friendly ping @TheCharlatan @real-or-random @theuni @fanquake  :)

**Comment by hebasto** [comment]
Rebased.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,21 @@
...
+#=============================
+# secp256k1 subtree
+#=============================
+set(SECP256K1_DISABLE_SHARED ON CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_ECDH OFF CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_RECOVERY ON CACHE BOOL "" FORCE)
+set(SECP256K1_ECMULT_GEN_KB 86 CACHE BOOL "" FORCE)
+include(GetTargetInterface)
+get_target_interface(core_c_flags "" core_base_interface COMPILE_OPTIONS)
+set(SECP256K1_LATE_CFLAGS ${core_c_flags} CACHE STRING "" FORCE)
```
Comment (real-or-random) [inline src/CMakeLists.txt:14]: What are the flags added on various platforms?

Comment (fanquake) [reply src/CMakeLists.txt:14]: > The core_c_flags variable holds the COMPILE_OPTIONS property of the core_base_interface library, which contains:
> There are no compile options that are specific to other platforms.

Why are we adding that flag specifically, into cflags, for mingw, and no other similar flags on different platforms?

Comment (fanquake) [reply src/CMakeLists.txt:14]: re offline discussion: why do we have a `core_c_flags` variable, if we aren't otherwise loading or using a c compiler (and how are things ending up in it).

Comment (real-or-random) [reply src/CMakeLists.txt:14]: - And these are cxx_flags, so they shouldn't apply to libsecp256k1? Or is it just that `try_append_cxx_flags` is a bit of a misleading name (in the end it's just COMPILE_OPTIONS?) I'm not sure after looking at https://github.com/hebasto/bitcoin/blob/cmake-staging/cmake/module/TryAppendCXXFlags.cmake .

 - Anyway, I wonder if this flag should be part of the `core_base_interface`. Just because it's useful for core, doesn't mean we should pass it also when building the dependencies.

Comment (real-or-random) [reply src/CMakeLists.txt:14]: > re offline discussion: why do we have a `core_c_flags` variable, if we aren't otherwise loading or using a c compiler (and how are things ending up in it).

Not just if this answers your question, but the variable is defined only in the `get_target_interface(core_c_flags ...`  It's just a local variable in this file and unset further below.

**Comment by hebasto** [comment]
The recent push makes the main build system pass only sanitizer options to the libsecp's build system, which mirrors the master branch: https://github.com/hebasto/bitcoin/blob/1b3b155fed4457f75df8cb5601b9a66d77e47662/configure.ac#L1746-L1750

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -1,6 +1,21 @@
...
+set(SECP256K1_ENABLE_MODULE_ECDH OFF CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_RECOVERY ON CACHE BOOL "" FORCE)
+set(SECP256K1_ECMULT_GEN_KB 86 CACHE BOOL "" FORCE)
+include(GetTargetInterface)
+get_target_interface(core_c_flags "" sanitize_interface COMPILE_OPTIONS)
+set(SECP256K1_LATE_CFLAGS ${core_c_flags} CACHE STRING "" FORCE)
+unset(core_c_flags)
+add_subdirectory(secp256k1 EXCLUDE_FROM_ALL)
+set_target_properties(secp256k1 PROPERTIES EXPORT_COMPILE_COMMANDS OFF)
+string(APPEND CMAKE_C_COMPILE_OBJECT " ${APPEND_CPPFLAGS} ${APPEND_CFLAGS}")
```
Comment (real-or-random) [inline src/CMakeLists.txt:18]: Not sure if this is the right thing, but I think it's orthogonal to this PR and should not hold it up. I've created https://github.com/hebasto/bitcoin/issues/239 to discuss this.

Comment (real-or-random) [inline src/CMakeLists.txt:18]: Not sure if this is the right thing, but I think it's orthogonal to this PR and should not hold it up. I've created https://github.com/hebasto/bitcoin/issues/239 to discuss this.

File: src/CMakeLists.txt (hunk 3 above)
Comment (hebasto) [reply src/CMakeLists.txt]: See also https://github.com/bitcoin-core/secp256k1/pull/1545#issuecomment-2182361968.

File: src/CMakeLists.txt (hunk 4 above)
Comment (hebasto) [reply src/CMakeLists.txt]: Thanks! Reworked.

**Comment by hebasto** [comment]
Addressed @real-or-random's https://github.com/bitcoin-core/secp256k1/pull/1543#issuecomment-2176914205:
> > Yeah, I tend to agree that we shouldn't touch these variables.
> 
> Hm, on a second thought, I'm not sure. I believe we want to build libsecp256k1 with the default `RelWithDebInfo` (`-O2 -g` in our case) even when we build it as part of Core. Because (almost) all our testing uses this config.
>
> edit: What does the current autotools build do?

Autotools builds with `-O2 -g`.

**Comment by hebasto** [comment]
Rebased.

**Comment by hebasto** [comment]
The last commit has been reworked to avoid hardcoded "RelWithDebInfo" configuration flags.

**Comment by fanquake** [comment]
> The last commit has been reworked to avoid hardcoded "RelWithDebInfo" configuration flags.

Is there any reason we don't want to hardcode the exact built type we want to build with? This isn't something that should change randomly, and seems like this has now just complicated the build logic?

**Comment by hebasto** [comment]
> > The last commit has been reworked to avoid hardcoded "RelWithDebInfo" configuration flags.
> 
> Is there any reason we don't want to hardcode the exact built type we want to build with? This isn't something that should change randomly, and seems like this has now just complicated the build logic?

The "RelWithDebInfo" is hardcoded. Its flags are not, which seems more maintainable.

**Comment by fanquake** [comment]
Ok, so after offline discussion, the above is only due to multi-config related things, which I've now discovered we are actually forced into supporting, because MSVC/windows CMake doesn't support anything single-config.. Seems like quite a lot of work has to be done / complication added to our build system just for this.

**Comment by hebasto** [comment]
> Seems like quite a lot of work has to be done / complication added to our build system just for this.

That's true. For more details, you can refer to CMake docs: https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#build-configurations

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -1,6 +1,32 @@
...
+unset(core_sanitizer_cxx_flags)
+# We want to build libsecp256k1 with the most tested RelWithDebInfo configuration.
+enable_language(C)
+foreach(config IN LISTS CMAKE_BUILD_TYPE CMAKE_CONFIGURATION_TYPES)
+  if(config STREQUAL "")
+    continue()
+  endif()
+  string(TOUPPER "${config}" config)
+  set(CMAKE_C_FLAGS_${config} "${CMAKE_C_FLAGS_RELWITHDEBINFO}")
+endforeach()
```
Comment (real-or-random) [inline src/CMakeLists.txt:26]: I agree this adds complexity, but I think it's currently the most uncontroversial version because it just replicates what autotools does. We may be able to remove this code after the discussion in #1549 (and just build as `Release`), but I'd like to have some more libsecp256k1 eyes on this. So, in the interest of not holding this PR up, Concept ACK on building as `RelWithDebInfo`.

Comment (hebasto) [reply src/CMakeLists.txt:26]: @real-or-random

You mean https://github.com/bitcoin-core/secp256k1/issues/1549, right?

Comment (real-or-random) [reply src/CMakeLists.txt:26]: oops, sure!

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -1,6 +1,32 @@
...
-# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+# file COPYING or https://opensource.org/license/mit/.
+
+#=============================
+# secp256k1 subtree
+#=============================
+set(SECP256K1_DISABLE_SHARED ON CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_ECDH OFF CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_RECOVERY ON CACHE BOOL "" FORCE)
+set(SECP256K1_ECMULT_GEN_KB 86 CACHE BOOL "" FORCE)
```
Comment (fanquake) [inline src/CMakeLists.txt:11]: Is there anything else we can/should be turning off here. i.e if we never intend to compile or run any tests, then we don't need to check for valgrind, or similar.

**Comment by fanquake** [comment]
Can the secp256k1 configuration be changed so that it is either the last thing to run before we print our own summary (same at autotools), or so that it's summary is printed just before ours, so that both summaries are actually printed at the end of configuring, rather than printing the secp256k1 summary, then running other checks / printing more output, and then printing our own summary.

**Comment by hebasto** [comment]
> Can the secp256k1 configuration be changed so that it is either the last thing to run before we print our own summary (same at autotools), or so that it's summary is printed just before ours, so that both summaries are actually printed at the end of configuring, rather than printing the secp256k1 summary, then running other checks / printing more output, and then printing our own summary.

Thanks! Done.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
+# file COPYING or https://opensource.org/license/mit/.
+
+#=============================
+# bitcoin_crypto library
+#=============================
+# This subdirectory includes compiler flag checks, which print
+# their results. Therefore, we place it before the secp256k1
+# subtree. This ensures that both summaries—secp256k1's and
+# Bitcoin Core's—are displayed one after the other.
+add_subdirectory(crypto)
```
Comment (fanquake) [inline src/CMakeLists.txt:12]: Looks like this shuffling broke most builds. Rather than moving this around, why can't we just do the secp256k1 configuration last?

File: src/CMakeLists.txt (hunk 7 above)
Comment (real-or-random) [reply src/CMakeLists.txt:28]: Autotools currently disables the benchmarks.
```suggestion
set(SECP256K1_BUILD_BENCHMARK OFF CACHE BOOL "" FORCE)
```

The tests are currently enabled. Are these run as part of any test or CI in Core? (I don't think so.) So we could also disable them.  `SECP256K1_BUILD_TESTS`, `SECP256K1_BUILD_EXHAUSTIVE_TESTS`, `SECP256K1_BUILD_CTIME_TESTS` 

(I may be a good idea to run the tests occasionally, e.g., when creating official builds. But that's an orthogonal issue.)

Comment (hebasto) [reply src/CMakeLists.txt:28]: > we don't need to check for valgrind

I've added setting `SECP256K1_VALGRIND`.

Speaking of `SECP256K1_BUILD_*` variables, `EXCLUDE_FROM_ALL` does this job and @theuni's [opinion](https://github.com/hebasto/bitcoin/pull/192#discussion_r1594428809) was:
> We could remove the `SECP256K1_BUILD_*` options and not have to worry about any new binaries being added in the future.

File: src/CMakeLists.txt (hunk 8 above)
Comment (hebasto) [reply src/CMakeLists.txt:18]: The `secp256k1` target, as a CMake's abstraction, must be known before its usage in statements like https://github.com/hebasto/bitcoin/blob/74bf342e05d409c800edae2f69fcea6a9c38a4e9/src/CMakeLists.txt#L94-L99

Comment (fanquake) [reply src/CMakeLists.txt:18]: By last, I mean:
```diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 18d07e56ef..d2ef9860d6 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -8,14 +8,35 @@ include(AddWindowsResources)
 configure_file(${CMAKE_SOURCE_DIR}/cmake/bitcoin-config.h.in config/bitcoin-config.h @ONLY)
 include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_CURRENT_SOURCE_DIR})
 
-#=============================
-# bitcoin_crypto library
-#=============================
-# This subdirectory includes compiler flag checks, which print
-# their results. Therefore, we place it before the secp256k1
-# subtree. This ensures that both summaries—secp256k1's and
-# Bitcoin Core's—are displayed one after the other.
+# After the transition from Autotools to CMake, the obj/ subdirectory
+# could be dropped as its only purpose was to separate a generated header
+# from source files.
+add_custom_target(generate_build_info
+  BYPRODUCTS ${PROJECT_BINARY_DIR}/src/obj/build.h
+  COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/src/obj
+  COMMAND ${CMAKE_COMMAND} -DBUILD_INFO_HEADER_PATH=${PROJECT_BINARY_DIR}/src/obj/build.h -DSOURCE_DIR=${PROJECT_SOURCE_DIR} -P ${CMAKE_SOURCE_DIR}/cmake/script/GenerateBuildInfo.cmake
+  COMMENT "Generating obj/build.h"
+  VERBATIM
+)
+add_library(bitcoin_clientversion OBJECT EXCLUDE_FROM_ALL
+  clientversion.cpp
+)
+target_compile_definitions(bitcoin_clientversion
+  PRIVATE
+    HAVE_BUILD_INFO
+)
+target_link_libraries(bitcoin_clientversion
+  PRIVATE
+    core_interface
+)
+add_dependencies(bitcoin_clientversion generate_build_info)
+
 add_subdirectory(crypto)
+add_subdirectory(univalue)
+add_subdirectory(util)
+if(WITH_MULTIPROCESS)
+  add_subdirectory(ipc)
+endif()
 
 #=============================
 # secp256k1 subtree
@@ -46,35 +67,6 @@ add_subdirectory(secp256k1 EXCLUDE_FROM_ALL)
 set_target_properties(secp256k1 PROPERTIES EXPORT_COMPILE_COMMANDS OFF)
 string(APPEND CMAKE_C_COMPILE_OBJECT " ${APPEND_CPPFLAGS} ${APPEND_CFLAGS}")
 
-# After the transition from Autotools to CMake, the obj/ subdirectory
-# could be dropped as its only purpose was to separate a generated header
-# from source files.
-add_custom_target(generate_build_info
-  BYPRODUCTS ${PROJECT_BINARY_DIR}/src/obj/build.h
-  COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/src/obj
-  COMMAND ${CMAKE_COMMAND} -DBUILD_INFO_HEADER_PATH=${PROJECT_BINARY_DIR}/src/obj/build.h -DSOURCE_DIR=${PROJECT_SOURCE_DIR} -P ${CMAKE_SOURCE_DIR}/cmake/script/GenerateBuildInfo.cmake
-  COMMENT "Generating obj/build.h"
-  VERBATIM
-)
-add_library(bitcoin_clientversion OBJECT EXCLUDE_FROM_ALL
-  clientversion.cpp
-)
-target_compile_definitions(bitcoin_clientversion
-  PRIVATE
-    HAVE_BUILD_INFO
-)
-target_link_libraries(bitcoin_clientversion
-  PRIVATE
-    core_interface
-)
-add_dependencies(bitcoin_clientversion generate_build_info)
-
-add_subdirectory(univalue)
-add_subdirectory(util)
-if(WITH_MULTIPROCESS)
-  add_subdirectory(ipc)
-endif()
-
 
 # Stable, backwards-compatible consensus functionality.
 add_library(bitcoin_consensus STATIC EXCLUDE_FROM_ALL
```

Comment (hebasto) [reply src/CMakeLists.txt]: Accepted.

File: src/CMakeLists.txt (hunk 7 above)
Comment (real-or-random) [reply src/CMakeLists.txt]: > > we don't need to check for valgrind
> 
> I've added setting `SECP256K1_VALGRIND`.
> 

Ah, sorry, I should have commented on this instead of just ignoring it.  Can you change it back?

This has a history: It is one specific instance of the discussion in https://github.com/bitcoin-core/secp256k1/issues/1549, where our decision (a few years ago) was to leave Valgrind enabled in production builds (if possible), to make sure that the `ctime_tests` binary with Valgrind on matches production builds as closely as possible, see https://github.com/bitcoin-core/secp256k1/pull/813#issuecomment-691720230 and the comment below it for detailed rationale.

---

> Speaking of `SECP256K1_BUILD_*` variables, `EXCLUDE_FROM_ALL` does this job and @theuni's [opinion](https://github.com/hebasto/bitcoin/pull/192#discussion_r1594428809) was:
> 
> > We could remove the `SECP256K1_BUILD_*` options and not have to worry about any new binaries being added in the future.

Oh sure, I forgot!

Comment (fanquake) [reply src/CMakeLists.txt]: > to make sure that the ctime_tests binary with Valgrind on matches production builds as closely as possible, see https://github.com/bitcoin-core/secp256k1/pull/813#issuecomment-691720230 and the comment below it for detailed rationale.

Does this still apply if we aren't (currently) actually building or running any tests (it's not super clear to me if this is the right decision)?

Comment (real-or-random) [reply src/CMakeLists.txt]: Yes, I think so. The library is written in a way such that the constant-time tests have a hidden switch to tell the library to do some Valgrind magic. This was really done with the idea that you can run the constant-time tests against the actual production library (you can literally link it to shared library) without any code changes (e.g., alignment, size, etc.) that could influence the compiler's decision to optimize or not optimize our constant-time hacks.  

So under the assumption that the same "misoptimization" happens with both the compiler used in production and the compiler used in libsecp256k1 testing, this is useful. (But sure, all of this would be even better if we ran the constant-time tests here because then the compilers would be the same and the assumption would hold trivially. Perhaps we should open an issue for this. Building and running the ctime_tests binary takes less than a second on my machine, so this should not be a burden for the test suite in Core.)

Does this make sense?

Comment (fanquake) [reply src/CMakeLists.txt]: > (But sure, all of this would be even better if we ran the constant-time tests here because then the compilers would be the same and the assumption would hold trivially. Perhaps we should open an issue for this.

Comment (fanquake) [reply src/CMakeLists.txt]: > (But sure, all of this would be even better if we ran the constant-time tests here because then the compilers would be the same and the assumption would hold trivially. Perhaps we should open an issue for this.

I guess this is fairly related to https://github.com/bitcoin-core/secp256k1/issues/1549, and actually running tests against code, when varying from the default secp256k1 configs, so maybe we can continue discussion there? I was planning no commenting soon, and think we should have some sort of agreement on what we are doing before moving ahead here without running any tests at all.

**Comment by hebasto** [comment]
Considering the recent discussion, which was nicely summarized in https://github.com/hebasto/bitcoin/pull/192#discussion_r1652971487, it seems reasonable to proceed in this PR with enabling the secp256k1 tests as it is done in the master branch.

Two new commits did this job.

File: src/CMakeLists.txt (hunk 7 above)
Comment (real-or-random) [reply src/CMakeLists.txt]: So I think `SECP256K1_VALGRIND` should be left on (=status quo in autotools) for now, at least until we come to a different conclusion in   https://github.com/bitcoin-core/secp256k1/issues/1549 .

Comment (hebasto) [reply src/CMakeLists.txt]: @real-or-random 
> So I think `SECP256K1_VALGRIND` should be left on (=status quo in autotools) for now, at least until we come to a different conclusion in [bitcoin-core/secp256k1#1549](https://github.com/bitcoin-core/secp256k1/issues/1549) .

`set(SECP256K1_VALGRIND ON)` will require valgrind installed, which is not always the case, at least in the Guix environment.

I suggest to skip setting `SECP256K1_VALGRIND` value from here.

Comment (real-or-random) [reply src/CMakeLists.txt]: > I suggest to skip setting `SECP256K1_VALGRIND` value from here.

Ok yes, that's what I meant, sorry.

**Comment by hebasto** [comment]
Addressed @real-or-random's https://github.com/hebasto/bitcoin/pull/192#discussion_r1658539148.

File: src/CMakeLists.txt (hunk 7 above)
Comment (hebasto) [reply src/CMakeLists.txt]: > > I suggest to skip setting `SECP256K1_VALGRIND` value from here.
> 
> Ok yes, that's what I meant, sorry.

Thanks! [Done](https://github.com/hebasto/bitcoin/pull/192#issuecomment-2196678927).

**Comment by hebasto** [comment]
Rebased to resolve a conflict.

**Comment by hebasto** [comment]
I've updated the PR descriptions with possible ways to test this PR.

File: src/CMakeLists.txt (hunk 8 above)
Comment (fanquake) [reply src/CMakeLists.txt]: Looks like this was broken again? Qt compiler checks and output appears after secp256k1 configure.

File: CMakeLists.txt [hunk 9]
```diff
@@ -599,6 +580,9 @@ if(DEFINED ENV{LDFLAGS})
   deduplicate_flags(CMAKE_EXE_LINKER_FLAGS)
 endif()
 
+if(BUILD_TESTS)
```
Comment (fanquake) [inline CMakeLists.txt:583]: Why is this needed for secp256k1 specifically?

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -38,6 +38,38 @@ if(WITH_MULTIPROCESS)
...
+set(SECP256K1_ENABLE_MODULE_ECDH OFF CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_RECOVERY ON CACHE BOOL "" FORCE)
+set(SECP256K1_ECMULT_GEN_KB 86 CACHE STRING "" FORCE)
+set(SECP256K1_BUILD_BENCHMARK OFF CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_TESTS ${BUILD_TESTS} CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_EXHAUSTIVE_TESTS ${BUILD_TESTS} CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_CTIME_TESTS OFF CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+include(GetTargetInterface)
+# -fsanitize and related flags apply to both C++ and C,
```
Comment (fanquake) [inline src/CMakeLists.txt:56]: Shouldn't this also be passing down LDFLAGS? Or is that being done somwhere else?

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -38,6 +38,38 @@ if(WITH_MULTIPROCESS)
...
+# so we can pass them down to libsecp256k1 as CFLAGS.
+get_target_interface(core_sanitizer_cxx_flags "" sanitize_interface COMPILE_OPTIONS)
+set(SECP256K1_LATE_CFLAGS ${core_sanitizer_cxx_flags} CACHE STRING "" FORCE)
+unset(core_sanitizer_cxx_flags)
+# We want to build libsecp256k1 with the most tested RelWithDebInfo configuration.
```
Comment (fanquake) [inline src/CMakeLists.txt:61]: > We want to build libsecp256k1 with the most tested RelWithDebInfo configuration.

Not sure this comment adds much value, and the CMake config is not really the most relevant thing? I'd say what matters more is the secp256k1 configuration, which currently, diverges from upstream, and isn't extensively tested.

File: CMakeLists.txt (hunk 9 above)
Comment (hebasto) [reply CMakeLists.txt:583]: It is not secp256k1-specific change. As a part of the "[FIXUP] cmake: Enable tests before adding subdirectories" commit this change adjusts the moment when `enable_testing()` being called. Previously, it was done implicitly via `include(CTest)` command _after_ processing the `src` subdirectory`. Now it done explicitly and _before_ processing the `src` subdirectory. Which activates any `add_test` commands within the `src` subdirectory including the `src/secp256k1` subdirectory as well.

File: src/CMakeLists.txt [hunk 12]
```diff
@@ -38,6 +38,38 @@ if(WITH_MULTIPROCESS)
...
+message("")
+message("Configuring secp256k1 subtree...")
+set(SECP256K1_DISABLE_SHARED ON CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_ECDH OFF CACHE BOOL "" FORCE)
+set(SECP256K1_ENABLE_MODULE_RECOVERY ON CACHE BOOL "" FORCE)
+set(SECP256K1_ECMULT_GEN_KB 86 CACHE STRING "" FORCE)
+set(SECP256K1_BUILD_BENCHMARK OFF CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_TESTS ${BUILD_TESTS} CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_EXHAUSTIVE_TESTS ${BUILD_TESTS} CACHE BOOL "" FORCE)
+set(SECP256K1_BUILD_CTIME_TESTS OFF CACHE BOOL "" FORCE)
```
Comment (fanquake) [inline src/CMakeLists.txt:53]: I thought we landed on that running the ctime tests would actually be valuable? If not, or if it's waiting on something else, it'd be good to have a comment here explaining why these tests are being skipped.

Comment (hebasto) [reply src/CMakeLists.txt:53]: I'm sorry if I misunderstood the previous discussion, but I thought that the idea was to mirror the master branch behaviour during migration to CMake.

File: src/CMakeLists.txt (hunk 11 above)
Comment (hebasto) [reply src/CMakeLists.txt:61]: Can you suggest an improved comment or other approach?

File: src/CMakeLists.txt (hunk 12 above)
Comment (fanquake) [reply src/CMakeLists.txt:53]: > I thought that the idea was to mirror the master branch behaviour during migration to CMake.

Ok, but the ctime tests are currently compiled, and runnable there.

Comment (hebasto) [reply src/CMakeLists.txt:53]: Thanks! Fixed in https://github.com/hebasto/bitcoin/pull/253.
